File: src/sdd_frl/findings_report.py
from __future__ import annotations

import re
from typing import Any

OUTCOME_LABELS = {
    "achieved": "已达成",
    "not_achieved": "未达成",
    "unknown": "未知",
}
OPTIMIZATION_LABELS = {
    "prompt": "Prompt",
    "skill": "Skill",
    "agent": "Agent",
    "unknown": "暂无法确定",
}


def _plain(value: Any) -> str:
    return re.sub(r"\s+", " ", str(value or "")).strip()


def _linked_alignments(
    task: dict[str, Any],
    divergence_id: str,
) -> list[dict[str, Any]]:
    return [
        item
        for item in task.get("alignments", [])
        if item.get("divergence_id") == divergence_id
    ]


def render_findings_section(
    findings: dict[str, Any],
    evidence: dict[str, Any] | None = None,
) -> str:
    """Render a human-readable task review without exposing raw evidence."""
    del evidence
    tasks = findings.get("task_episodes", [])
    lines = ["## 任务分析", ""]
    if not tasks:
        lines.append("本周期没有形成可分析任务。")
        return "\n".join(lines)

    all_divergences: list[dict[str, Any]] = []
    for index, task in enumerate(tasks, start=1):
        lines.extend([
            f"### {index}. {_plain(task.get('goal')) or '未命名任务'}",
            "",
            f"**结果：{OUTCOME_LABELS.get(task.get('outcome_status'), '未知')}**",
            "",
            "**用户目标**",
            "",
            _plain(task.get("expected_outcome")) or "未记录明确预期。",
            "",
            "**过程**",
            "",
        ])
        summaries = [
            _plain(item) for item in task.get("execution_summary", []) if _plain(item)
        ]
        lines.extend(f"- {item}" for item in summaries)
        if not summaries:
            lines.append("- 未记录关键过程。")

        lines.extend(["", "**分歧与对齐**", ""])
        divergences = task.get("divergences", [])
        if not divergences:
            lines.append("未发现有证据支持的分歧。")
        for divergence in divergences:
            all_divergences.append(divergence)
            status = "已解决" if divergence["status"] == "resolved" else "未解决"
            lines.extend([
                f"**分歧：{_plain(divergence['summary'])}（{status}）**",
                "",
                f"- 用户期望：{_plain(divergence['user_expectation'])}",
                f"- Agent 实际行为：{_plain(divergence['agent_behavior'])}",
            ])
            alignments = _linked_alignments(task, divergence["divergence_id"])
            if alignments:
                lines.append(
                    "- 对齐结果："
                    + "；".join(_plain(item["summary"]) for item in alignments)
                )
                lines.append(
                    "- 后续动作："
                    + "；".join(_plain(item["resulting_action"]) for item in alignments)
                )
            else:
                lines.append("- 对齐结果：尚未完成对齐。")
            lines.extend([
                f"- 根因：{_plain(divergence['root_cause'])}",
                (
                    "- 优化对象："
                    f"{OPTIMIZATION_LABELS.get(divergence['optimization_target'], '暂无法确定')}"
                ),
                f"- 优化方向：{_plain(divergence['optimization_direction'])}",
                f"- 验收方式：{_plain(divergence['acceptance_check'])}",
                "",
            ])

        independent_alignments = [
            item
            for item in task.get("alignments", [])
            if item.get("divergence_id") is None
        ]
        for alignment in independent_alignments:
            lines.extend([
                f"- 对齐：{_plain(alignment['summary'])}",
                f"- 后续动作：{_plain(alignment['resulting_action'])}",
            ])
        if lines[-1] != "":
            lines.append("")

    lines.extend(["## 优化清单", ""])
    if not all_divergences:
        lines.append("本周期没有发现需要修改 Prompt、Skill 或 Agent 的分歧。")
    else:
        for divergence in sorted(
            all_divergences,
            key=lambda item: item["status"] == "resolved",
        ):
            target = OPTIMIZATION_LABELS.get(
                divergence["optimization_target"],
                "暂无法确定",
            )
            lines.append(
                f"- **{target}**：{_plain(divergence['optimization_direction'])}"
                f"（{_plain(divergence['summary'])}）"
            )
    return "\n".join(lines).rstrip()

File: src/findings-report.js
const OUTCOME_LABELS = {
  achieved: "已达成",
  not_achieved: "未达成",
  unknown: "未知"
};

const OPTIMIZATION_LABELS = {
  prompt: "Prompt",
  skill: "Skill",
  agent: "Agent",
  unknown: "暂无法确定"
};

function plain(value) {
  return String(value ?? "").replace(/\s+/g, " ").trim();
}

function linkedAlignments(task, divergenceId) {
  return (task.alignments ?? []).filter(
    (item) => item.divergence_id === divergenceId
  );
}

export function renderFindingsSection(findings, evidence = null) {
  void evidence;
  const tasks = findings.task_episodes ?? [];
  const lines = ["## 任务分析", ""];
  if (!tasks.length) {
    lines.push("本周期没有形成可分析任务。");
    return lines.join("\n");
  }

  const allDivergences = [];
  for (const [index, task] of tasks.entries()) {
    lines.push(
      `### ${index + 1}. ${plain(task.goal) || "未命名任务"}`,
      "",
      `**结果：${OUTCOME_LABELS[task.outcome_status] ?? "未知"}**`,
      "",
      "**用户目标**",
      "",
      plain(task.expected_outcome) || "未记录明确预期。",
      "",
      "**过程**",
      ""
    );
    const summaries = (task.execution_summary ?? []).map(plain).filter(Boolean);
    if (summaries.length) lines.push(...summaries.map((item) => `- ${item}`));
    else lines.push("- 未记录关键过程。");

    lines.push("", "**分歧与对齐**", "");
    const divergences = task.divergences ?? [];
    if (!divergences.length) lines.push("未发现有证据支持的分歧。");
    for (const divergence of divergences) {
      allDivergences.push(divergence);
      const status = divergence.status === "resolved" ? "已解决" : "未解决";
      lines.push(
        `**分歧：${plain(divergence.summary)}（${status}）**`,
        "",
        `- 用户期望：${plain(divergence.user_expectation)}`,
        `- Agent 实际行为：${plain(divergence.agent_behavior)}`
      );
      const alignments = linkedAlignments(task, divergence.divergence_id);
      if (alignments.length) {
        lines.push(
          `- 对齐结果：${alignments.map((item) => plain(item.summary)).join("；")}`,
          `- 后续动作：${alignments.map((item) => plain(item.resulting_action)).join("；")}`
        );
      } else {
        lines.push("- 对齐结果：尚未完成对齐。");
      }
      lines.push(
        `- 根因：${plain(divergence.root_cause)}`,
        `- 优化对象：${OPTIMIZATION_LABELS[divergence.optimization_target] ?? "暂无法确定"}`,
        `- 优化方向：${plain(divergence.optimization_direction)}`,
        `- 验收方式：${plain(divergence.acceptance_check)}`,
        ""
      );
    }

    const independentAlignments = (task.alignments ?? []).filter(
      (item) => item.divergence_id === null
    );
    for (const alignment of independentAlignments) {
      lines.push(
        `- 对齐：${plain(alignment.summary)}`,
        `- 后续动作：${plain(alignment.resulting_action)}`
      );
    }
    if (lines.at(-1) !== "") lines.push("");
  }

  lines.push("## 优化清单", "");
  if (!allDivergences.length) {
    lines.push("本周期没有发现需要修改 Prompt、Skill 或 Agent 的分歧。");
  } else {
    allDivergences
      .sort((left, right) => Number(left.status === "resolved") - Number(right.status === "resolved"))
      .forEach((divergence) => {
        const target = OPTIMIZATION_LABELS[divergence.optimization_target] ?? "暂无法确定";
        lines.push(
          `- **${target}**：${plain(divergence.optimization_direction)}` +
          `（${plain(divergence.summary)}）`
        );
      });
  }
  return lines.join("\n").trimEnd();
}
